API Client that must be handled by implementing gRPC API Server with business logic

LinMAD opened the pull request:
Some instructions about test assignment and bare Golang code for API Client that generates requests to the server.

Things to cover
- [x] Review Instructions if they clear
- [x] CI
- [x] Docker

LinMAD committed:
WIP Add base idea with structure and instructions

LinMAD committed:
Add test client API that must be handled in the test assignment

LinMAD committed:
Add simple result table with exec time to compare with

LinMAD committed:
Dokerize application

LinMAD committed:
Update CI to build project

LinMAD committed:
Fix License information

LinMAD committed:
Fix lint CI to validate all codebase - resolving issues with missing context

LinMAD committed:
Update dependabot config

LinMAD committed:
Update golang-ci configuration for linting

LinMAD committed:
Remove not used for public workflow and clean up irrelevant code ownership

LinMAD committed:
Update cache for golang

LinMAD committed:
Update CI

LinMAD committed:
Update catalog configuration for techdocs

LinMAD committed:
Update techdocs

nhhagen commented on .vale.ini:
```diff
@@ -0,0 +1,9 @@
+
+Packages = https://github.com/coopnorge/vale-coop/releases/latest/download/Coop.zip
+
+Vocab = General, CoopAbbreviations, CoopGeneral
```
These are loaded automatically by the package

```suggestion
```

nhhagen commented on .vale.ini:
```diff
@@ -0,0 +1,9 @@
+
+Vocab = General, CoopAbbreviations, CoopGeneral
+
+[*.md]
+TokenIgnores = [\w][\w.]*@[\w*][\w.]*[\w]
+BasedOnStyles = Vale
```
This is also activated by the loaded package.

```suggestion
```

LinMAD committed:
Add example of Coop Logistics output

nhhagen commented on api/README.md:
```diff
@@ -0,0 +1,21 @@
+- [Go](https://go.dev/doc/install)
+- [Buf for proto-files management](https://buf.build/docs/installation)
+
+Check make file in root of the project for compiling go code from proto-files.
```
```suggestion
Check the `Makefile` in root of the project for compiling go code from proto-files.
```

LinMAD committed:
Fix links to the sources in instructions file

LinMAD committed:
nhhagen suggestion Update .vale.ini

Co-authored-by: Niels Henrik Hagen <nhhagen@gmail.com>

LinMAD committed:
nhhagen suggestion Update .vale.ini

Co-authored-by: Niels Henrik Hagen <nhhagen@gmail.com>

LinMAD committed:
nhhagen suggestion Update api/README.md

Co-authored-by: Niels Henrik Hagen <nhhagen@gmail.com>

nhhagen commented on mkdocs.yml:
`mkdocs.yml` should not be committed and added to `.gitignore`

nhhagen commented on docker-compose.yml:
```diff
@@ -0,0 +1,27 @@
+  techdocs:
+    build:
+      context: .
+      dockerfile: internal/techdocs/Dockerfile
```
Why are we not using our common pattern of having this in `/docker-compose/`?

LinMAD replied:
Minimized "Coop specifics" so for participants would be a bit easier, could create questions like "why docker-compose folder with docker images? etc"

We can move it to the root as well, not a big deal.

LinMAD committed:
Fix link to docker-compose file where golang app executed as service

LinMAD committed:
Add to gitignore mkdocs.yml

LinMAD committed:
Move techdocs docker image to docker compose folder to follow Coop common pattern

yurydavydov commented on docs/INDEX.md:
```diff
@@ -0,0 +1,57 @@
+
+## Acceptance Criteria
+
+- All three parts must be done (One, Two, Three).
```
You could delete the part in braces. It does not add any additional information.

yurydavydov commented on docs/INDEX.md:
```diff
@@ -0,0 +1,57 @@
+
+## Assignment Appendix
+
+Coop Logistics Engine, simulates several cargo units that
```
"," after "Engine" is unnecessary.

yurydavydov commented on docs/INDEX.md:
```diff
@@ -0,0 +1,57 @@
+checkout service in
+[docker-compose](../docker-compose.yml) "interview_backend_coop_logistics"
+
+## Assignment Details
```
Instead of "Assignment Details", I would call it "Assignment".

Further, I would rephrase:
"Develop an API system that matches the criteria described in parts. Here is [Swagger documentation with descriptions.](../api/logistics.swagger.json)"

to

"The result of this task should be an API system that fulfils the criteria described in three parts below. Here is the link to the API documentation [Swagger documentation with descriptions.](../api/logistics.swagger.json)"

LinMAD replied:
Nice, also added a link to protifle

yurydavydov commented on docs/INDEX.md:
```diff
@@ -0,0 +1,57 @@
+Develop an API system that matches the criteria described in parts,
+here is [Swagger documentation with descriptions.](../api/logistics.swagger.json)
+
+Don't hesitate to ask questions for clarification if you have them.
```
This "Call for action" and wishes good luck I would place at the end after part three.

LinMAD replied:
Maybe, but that feels like it will be a bit detached from the assignment task, where parts are pure instructions without distraction ;)

LinMAD replied:
Updated public part of that, internal code was not changed :)

yurydavydov commented on docs/INDEX.md:
```diff
@@ -0,0 +1,57 @@
+
+### Part One: Implement Backend API
+
+Your first task is to implement API server
```
I would rephrase it as "Implement the API server for that..."

1. You do not need to refer to the candidate "Your...". It's clear from the context who has to work on the task.
2. It's already in part one, so you do not have to tell "first". Keep it short.

yurydavydov commented on docs/INDEX.md:
```diff
@@ -0,0 +1,57 @@
+
+The solution should output a log message to STDOUT once per second with:
+
+- Number of received messages / second.
```
A list with just one item does not look good. Rephrase to "The solution should output a log message to STDOUT once per second with the number of received messages per second." Same in the part two.

LinMAD committed:
Update review notes related to assignment instructions

nhhagen commented on .github/workflows/build.yaml:
```diff
@@ -1,16 +1,39 @@
+    branches:
+      - main
   pull_request: {}
-concurrency:
-  group: ${{ github.workflow }}-${{ github.ref }}
-  cancel-in-progress: true
```
We should keep this so that we cancel duplicate builds

ezorrio commented on api/v1/logistics.proto:
```diff
@@ -0,0 +1,79 @@
+
+// Location where entity now located in X,Y Axis
+message Location {
+    int64 x = 1;
```
might be uint32/uint64, based on generator code

LinMAD replied:
Yeah, random generator anyway using only absolute values

ezorrio commented on docs/INDEX.md:
```diff
@@ -0,0 +1,60 @@
+# Interview Assignment - Coop Logistics API
```
As discussed, would be great to mention whether to implement client or server part, as we don't want to leave a room for different interpretation

stephendrake commented on api/README.md:
```diff
@@ -0,0 +1,21 @@
+# Contribute instructions
+
+Make sure you have before:
```
Perhaps:
```suggestion
Prerequisites:
```

stephendrake commented on docs/INDEX.md:
```diff
@@ -0,0 +1,63 @@
+## Acceptance Criteria
+
+- All three parts must be done.
+- API must be sustainable to handle requests without dropping them.
```
```suggestion
- API must be capable of handling all requests sent by this client without dropping them.
```

stephendrake commented on docs/INDEX.md:
```diff
@@ -0,0 +1,63 @@
+The result of this task should be an API system that fulfils the criteria
+described in three parts below. Here is the link to the API documentation
+[Swagger documentation](../api/logistics.swagger.json)
+or [Protofile](../api/v1/logistics.proto) with descriptions.
```
Minor: This line about the API specs is probably better suited in the "Part One" section, but as it's already mentioned there so one could probably remove it here.

Also, as the client expects a gRPC endpoint, so it's perhaps not worth mentioning the OpenAPI v2 spec until it's required.

LinMAD replied:
Yeah let's do it like that, when we will require to implement it as HTTP server like REST then we can return it back. Will refactor it to be more generic, like here is a folder with API specs.

stephendrake commented on docs/INDEX.md:
```diff
@@ -0,0 +1,63 @@
+
+### Part One: Implement Backend API
+
+Implement the API server for that
+[proto-file](../api/v1/logistics.proto).
```
```suggestion
Implement an API sever that provides the following service over gRPC: 
[proto-file](../api/v1/logistics.proto).
```

stephendrake commented on api/v1/logistics.proto:
```diff
@@ -0,0 +1,79 @@
+
+// Location where entity now located in X,Y Axis
+message Location {
+    uint32 x = 1;
```
Probably not that important as it's for an assessment, but does one not generally use latitude and longitude in logistics?

LinMAD replied:
@stephendrake We can but for simplicity, it's just an imaginary X,Y-Axis chart for immersive experience we can rename it :D

stephendrake replied:
Up to you, but if anything then just to avoid the inevitable "why no lat/long" that I would ask if I was doing the interview.

LinMAD committed:
Suggestion from  stephendrake - Update docs/INDEX.md

Co-authored-by: Stephen Drake <405056+stephendrake@users.noreply.github.com>

LinMAD committed:
Suggestion  from stephendrake - Update api/README.md

Co-authored-by: Stephen Drake <405056+stephendrake@users.noreply.github.com>

LinMAD committed:
Suggestion from  stephendrake - Update docs/INDEX.md

Co-authored-by: Stephen Drake <405056+stephendrake@users.noreply.github.com>

LinMAD committed:
Remove OpenAPI support and docs while assignment does not support it

LinMAD committed:
Rename properties in Location model to Latitude and Longitude

stephendrake commented on docs/INDEX.md:
```diff
@@ -0,0 +1,61 @@
+data comes into the API you implemented in *Part One*.
+Write at least one unit test that verifies that your solution works.
+
+The solution should output a log message to STDOUT once per second with:
```
```suggestion
```
Perhaps just rework this bit for clarity. Either this or the line below and subsequent once-per-second requirements are are duplicated.

stephendrake commented on README.md:
```diff
@@ -1,41 +1,23 @@
+-> [instructions](docs/INDEX.md).
 
-## Initialization instructions:
+Example of output:
```
```suggestion
Example of client program output:
```

stephendrake reviewed (commented):
A couple points I missed before.

LinMAD committed:
Update docker files and related code to allow set host and port for gRPC client

LinMAD committed:
Improved ASCII printer to align based on the lenght of the values in each column

LinMAD committed:
Add suggestion form stephendrake - Update docs/INDEX.md

Co-authored-by: Stephen Drake <405056+stephendrake@users.noreply.github.com>

LinMAD committed:
Suggestion form stephendrake - Update README.md

Co-authored-by: Stephen Drake <405056+stephendrake@users.noreply.github.com>

LinMAD commented:
@stephendrake Just a few bits that I added:
- Config support
- Updated docker-compse - now it will boost up setup since there is now a solution folder and docker-compse.yml preconfigured
- Minor ASCII printer now aligns rows length to values (pretty print)

LinMAD commented on docker-compose.yml:
```diff
@@ -0,0 +1,49 @@
+    container_name: interview_backend_coop_logistics_client
+    environment:
+      - CLIENT_SERVICE_PORT=50051
+      - CLIENT_SERVICE_HOST=interview_backend_coop_logistics_server
```
Yeah maybe I rush it a bit, let me know if the naming is confusing :D

LinMAD commented:
@nhhagen Any other contributions or are we good to go?

nhhagen commented:
Looks good, I focused on the case description for the candidate. Do we have people to test this?

LinMAD commented:
> Looks good, I focused on the case description for the candidate. Do we have people to test this?

We can speak about that in another thread :)
